### DISCUSSION TITLE
ESP32-S2/S3 and USB MIDI

### DISCUSSION THREAD

**Discussion by thijstriemstra** [corrados/edrumulus#38] [General]
Looks like the [ESP32-S2](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/hw-reference/esp32s2/user-guide-devkitm-1-v1.html) can also be used as a USB MIDI device(?), like Teensy can.

![esp32-s2 pin layout](https://user-images.githubusercontent.com/305679/148655723-bba8473c-da0b-4075-ba78-1818aadfb381.jpg)

Good detailed video of it can be found here: https://www.youtube.com/watch?v=XnqDEV21BSA

Looks like it uses [EspTinyUSB](https://github.com/chegewara/EspTinyUSB). They have a MIDI example here: https://github.com/chegewara/EspTinyUSB/tree/master/examples/device/midi

More about it on hackster.io: https://www.hackster.io/news/create-custom-usb-devices-with-the-esp32s2stick-458bf370a522

[TinyUSB](https://github.com/hathach/tinyusb) also mentions there's support for the ESP32-S3.

Hopefully this will allow us to skip Hairless MIDI when using ESP32(-S2/S3)?

**Comment by thijstriemstra** [comment]
Turns out ESP32-S2 is single-core, most likely not powerful enough for Edrumulus. The [ESP32-S3](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/hw-reference/esp32s3/user-guide-devkitc-1.html) is dual-core though and looks like the S2's big brother.

ESP32-S3 is [not cheap](https://www.aliexpress.com/item/1005003488961304.html) though (yet) at 16,70 euro for a board. So compared to Teensy 4.x it's maybe not a good alternative yet (as teensy is little bit more expensive but much faster).

![ESP32-S3-DevKitC-1-pin-layout](https://user-images.githubusercontent.com/305679/148692171-a356cac0-f8cc-4040-9ead-b44703552f59.png)

**Comment by thijstriemstra** [comment]
@corrados any thoughts on the S3 being able to handle both processing and midi usb, do you think it'll be fast enough?

**Reply by corrados** [reply]
If it is a dual-core 240 MHz then it should have comparable speed as the ESP32 you use right now, I assume.

**Comment by leendalex1993** [comment]
Nice

**Comment by thijstriemstra** [comment]
What specific ESP32-S3 board are you [going to test](https://github.com/corrados/edrumulus/discussions/72#discussioncomment-5497621) with @corrados? I'll order the same for testing. And will edrumulus support usb midi when s3 is implemented/tested?

**Reply by corrados** [reply]
> What specific ESP32-S3 board are you https://github.com/corrados/edrumulus/discussions/72#discussioncomment-5497621 with @corrados?

I have ordered this one: https://de.aliexpress.com/item/1005004637451669.html

> And will edrumulus support usb midi when s3 is implemented/tested?

Good question. I hope it will work but I cannot guarantee...

**Reply by thijstriemstra** [reply]
> Good question. I hope it will work but I cannot guarantee...

I should've phrased it differently. I'm glad to hear it's on the roadmap!

**Comment by davy39** [comment]
ESP32-S3 can be use with usb midi with tinyUSB with this PR #136 . 
It seems to work for me without problem, feel free to try it as well :)

**Reply by thijstriemstra** [reply]
> It seems to work for me without problem, feel free to try it as well :)

That's awesome! What esp32-s3 board did you test it with?

**Reply by davy39** [reply]
Both esp32-s3-devkitc-1 and Lolin S3 seems to work ok.
